File: frontend/src/shared/api/clash.ts
import { apiRequest } from '@/shared/api/client';
import type { ApiDecoder } from '@/shared/api/decoder';

export interface SourceConfig {
  id: number;
  source_type: 'upload' | 'subscription_url';
  source_url: string;
  content_type: string;
  fetched_at?: string;
  filename: string;
  content_hash: string;
  status: 'parsed' | 'imported';
  total_nodes: number;
  valid_nodes: number;
  invalid_nodes: number;
  duplicate_nodes: number;
  imported_nodes: number;
  uploaded_by: string;
  uploaded_by_id: number;
  created_at: string;
  updated_at: string;
}

export interface ParseIssue {
  index: number;
  name?: string;
  message: string;
}

export interface ParsedNode {
  name: string;
  type: string;
  server: string;
  port: number;
  fingerprint: string;
  metadata_json: string;
}

export interface ParseResult {
  nodes: ParsedNode[];
  issues: ParseIssue[];
}

export interface UploadSourceConfigResponse {
  config: SourceConfig;
  parse_result: ParseResult;
}

export interface ProxyNode {
  id: number;
  source_config_id: number;
  source_config_name: string;
  name: string;
  type: string;
  server: string;
  port: number;
  tags: string;
  metadata_json: string;
  enabled: boolean;
  last_test_status: 'unknown' | 'success' | 'failed';
  last_latency_ms?: number;
  last_test_error?: string;
  last_tested_at?: string;
  created_at: string;
  updated_at: string;
}

export interface TestNodeResult {
  node_id: number;
  node_name: string;
  success: boolean;
  latency_ms: number;
  error_message: string;
}

export interface PortProfileProxySettings {
  strategy_type: 'select' | 'url-test' | 'fallback' | 'load-balance';
  test_url: string;
  test_interval_seconds: number;
  load_balance_strategy: 'consistent-hashing' | 'round-robin';
  load_balance_lazy?: boolean;
  load_balance_disable_udp?: boolean;
  udp_enabled: boolean;
  auth_enabled: boolean;
  auth_username?: string;
  auth_password?: string;
}

export interface PortProfile {
  id: number;
  name: string;
  listen_host: string;
  mixed_port: number;
  socks_port: number;
  http_port: number;
  proxy_settings: PortProfileProxySettings;
  include_in_runtime: boolean;
  kernel_type: string;
  created_at: string;
  updated_at: string;
}

export interface RuntimeConfig {
  id: number;
  port_profile_id: number;
  kernel_type: string;
  checksum: string;
  rendered_config: string;
  created_at: string;
  updated_at: string;
}

export interface PortProfileWithNodes {
  profile: PortProfile;
  node_ids: number[];
  nodes?: ProxyNode[];
  runtime_config?: RuntimeConfig;
}

export interface PortProfileTemplate {
  id: number;
  name: string;
  description: string;
  mixed_port: number;
  proxy_settings: PortProfileProxySettings;
  created_at: string;
  updated_at: string;
}

export interface KernelInstance {
  id: number;
  kernel_type: string;
  status:
    | 'stopped'
    | 'starting'
    | 'running'
    | 'stopping'
    | 'error'
    | 'reloading';
  pid?: number;
  work_dir: string;
  config_path: string;
  controller_address: string;
  active_config_checksum: string;
  active_profile_count: number;
  active_listener_count: number;
  last_action: string;
  last_error: string;
  last_started_at?: string;
  last_stopped_at?: string;
  last_reloaded_at?: string;
  created_at: string;
  updated_at: string;
}

export interface RuntimeStatusView {
  kernel_instance: KernelInstance;
  is_process_alive: boolean;
  controller_ok: boolean;
  mihomo_version: string;
  active_profiles: PortProfileWithNodes[];
}

export interface KernelCapability {
  kernel_type: string;
  supports_reload: boolean;
  supports_external_api: boolean;
  supports_health_check: boolean;
  supported_strategies: string[];
}

export interface RenderResult {
  kernel_type: string;
  checksum: string;
  content: string;
}

function decoder<T>(): ApiDecoder<T> {
  return (val: unknown) => val as T;
}

// --- API Methods ---

export async function uploadSourceConfig(
  filename: string,
  rawContent: string,
): Promise<UploadSourceConfigResponse> {
  return apiRequest<UploadSourceConfigResponse>(
    '/api/v1/admin/clash/source-configs/upload',
    {
      method: 'POST',
      body: { filename, raw_content: rawContent },
    },
    decoder<UploadSourceConfigResponse>(),
  );
}

export async function fetchSubscription(
  sourceUrl: string,
): Promise<UploadSourceConfigResponse> {
  return apiRequest<UploadSourceConfigResponse>(
    '/api/v1/admin/clash/source-configs/subscription',
    {
      method: 'POST',
      body: { source_url: sourceUrl },
    },
    decoder<UploadSourceConfigResponse>(),
  );
}

export async function listSourceConfigs(
  page = 1,
  pageSize = 20,
): Promise<{
  items: SourceConfig[];
  total: number;
  page: number;
  page_size: number;
}> {
  return apiRequest<{
    items: SourceConfig[];
    total: number;
    page: number;
    page_size: number;
  }>(
    `/api/v1/admin/clash/source-configs?page=${page}&page_size=${pageSize}`,
    {
      method: 'GET',
    },
    decoder<{
      items: SourceConfig[];
      total: number;
      page: number;
      page_size: number;
    }>(),
  );
}

export async function deleteSourceConfig(
  id: number,
): Promise<{ deleted: boolean }> {
  return apiRequest<{ deleted: boolean }>(
    `/api/v1/admin/clash/source-configs/${id}`,
    {
      method: 'DELETE',
    },
    decoder<{ deleted: boolean }>(),
  );
}

export async function confirmSourceConfig(
  id: number,
): Promise<{ imported_nodes: number }> {
  return apiRequest<{ imported_nodes: number }>(
    `/api/v1/admin/clash/source-configs/${id}/confirm`,
    {
      method: 'POST',
    },
    decoder<{ imported_nodes: number }>(),
  );
}

export interface SyncSourceResult {
  id: number;
  total_nodes: number;
  valid_nodes: number;
  invalid_nodes: number;
  duplicate_nodes: number;
  imported_nodes: number;
}

export async function refreshSourceConfig(
  id: number,
): Promise<SyncSourceResult> {
  return apiRequest<SyncSourceResult>(
    `/api/v1/admin/clash/source-configs/${id}/refresh`,
    {
      method: 'POST',
    },
    decoder<SyncSourceResult>(),
  );
}

export async function reuploadSourceConfig(
  id: number,
  filename: string,
  rawContent: string,
): Promise<SyncSourceResult> {
  return apiRequest<SyncSourceResult>(
    `/api/v1/admin/clash/source-configs/${id}/reupload`,
    {
      method: 'POST',
      body: { filename, raw_content: rawContent },
    },
    decoder<SyncSourceResult>(),
  );
}

export async function listProxyNodes(params: {
  page?: number;
  pageSize?: number;
  keyword?: string;
  sourceConfigId?: number;
  enabled?: boolean;
}): Promise<{
  items: ProxyNode[];
  total: number;
  page: number;
  page_size: number;
}> {
  const query = new URLSearchParams();
  if (params.page) query.set('page', params.page.toString());
  if (params.pageSize) query.set('page_size', params.pageSize.toString());
  if (params.keyword) query.set('keyword', params.keyword);
  if (params.sourceConfigId)
    query.set('source_config_id', params.sourceConfigId.toString());
  if (params.enabled !== undefined)
    query.set('enabled', params.enabled ? 'true' : 'false');
  return apiRequest<{
    items: ProxyNode[];
    total: number;
    page: number;
    page_size: number;
  }>(
    `/api/v1/admin/clash/nodes?${query.toString()}`,
    {
      method: 'GET',
    },
    decoder<{
      items: ProxyNode[];
      total: number;
      page: number;
      page_size: number;
    }>(),
  );
}

export async function testProxyNodes(
  nodeIds: number[],
  binaryPath?: string,
  testUrl?: string,
  timeoutSeconds?: number,
): Promise<{ results: TestNodeResult[] }> {
  return apiRequest<{ results: TestNodeResult[] }>(
    '/api/v1/admin/clash/nodes/test',
    {
      method: 'POST',
      body: {
        node_ids: nodeIds,
        binary_path: binaryPath,
        test_url: testUrl,
        timeout_seconds: timeoutSeconds,
      },
    },
    decoder<{ results: TestNodeResult[] }>(),
  );
}

export async function deleteProxyNode(
  id: number,
): Promise<{ deleted: boolean }> {
  return apiRequest<{ deleted: boolean }>(
    `/api/v1/admin/clash/nodes/${id}`,
    {
      method: 'DELETE',
    },
    decoder<{ deleted: boolean }>(),
  );
}

export async function deleteProxyNodesBatch(
  ids: number[],
): Promise<{ deleted_count: number }> {
  return apiRequest<{ deleted_count: number }>(
    '/api/v1/admin/clash/nodes/batch-delete',
    {
      method: 'POST',
      body: { ids },
    },
    decoder<{ deleted_count: number }>(),
  );
}

export async function toggleProxyNodesBatch(
  ids: number[],
  enabled: boolean,
): Promise<{ toggled_count: number; enabled: boolean }> {
  return apiRequest<{ toggled_count: number; enabled: boolean }>(
    '/api/v1/admin/clash/nodes/batch-toggle',
    {
      method: 'POST',
      body: { ids, enabled },
    },
    decoder<{ toggled_count: number; enabled: boolean }>(),
  );
}

export async function listPortProfiles(): Promise<{
  items: PortProfileWithNodes[];
}> {
  return apiRequest<{ items: PortProfileWithNodes[] }>(
    '/api/v1/admin/clash/port-profiles',
    {
      method: 'GET',
    },
    decoder<{ items: PortProfileWithNodes[] }>(),
  );
}

export async function createPortProfile(payload: {
  name: string;
  listen_host?: string;
  mixed_port: number;
  socks_port?: number;
  http_port?: number;
  proxy_settings: PortProfileProxySettings;
  include_in_runtime?: boolean;
  node_ids?: number[];
}): Promise<PortProfileWithNodes> {
  return apiRequest<PortProfileWithNodes>(
    '/api/v1/admin/clash/port-profiles',
    {
      method: 'POST',
      body: payload,
    },
    decoder<PortProfileWithNodes>(),
  );
}

export async function updatePortProfile(
  id: number,
  payload: {
    name: string;
    listen_host?: string;
    mixed_port: number;
    socks_port?: number;
    http_port?: number;
    proxy_settings: PortProfileProxySettings;
    include_in_runtime?: boolean;
    node_ids?: number[];
  },
): Promise<PortProfileWithNodes> {
  return apiRequest<PortProfileWithNodes>(
    `/api/v1/admin/clash/port-profiles/${id}`,
    {
      method: 'PUT',
      body: payload,
    },
    decoder<PortProfileWithNodes>(),
  );
}

export async function deletePortProfile(
  id: number,
): Promise<{ deleted: boolean }> {
  return apiRequest<{ deleted: boolean }>(
    `/api/v1/admin/clash/port-profiles/${id}`,
    {
      method: 'DELETE',
    },
    decoder<{ deleted: boolean }>(),
  );
}

export async function previewPortProfile(id: number): Promise<RenderResult> {
  return apiRequest<RenderResult>(
    `/api/v1/admin/clash/port-profiles/${id}/preview`,
    {
      method: 'GET',
    },
    decoder<RenderResult>(),
  );
}

export async function listPortProfileTemplates(): Promise<{
  items: PortProfileTemplate[];
}> {
  return apiRequest<{ items: PortProfileTemplate[] }>(
    '/api/v1/admin/clash/port-profile-templates',
    {
      method: 'GET',
    },
    decoder<{ items: PortProfileTemplate[] }>(),
  );
}

export async function getRuntimeStatus(
  kernelType = 'mihomo',
): Promise<RuntimeStatusView> {
  return apiRequest<RuntimeStatusView>(
    `/api/v1/admin/clash/runtime/status?kernel_type=${kernelType}`,
    {
      method: 'GET',
    },
    decoder<RuntimeStatusView>(),
  );
}

export async function startKernel(payload: {
  kernel_type?: string;
  binary_path?: string;
  work_dir?: string;
  allow_lan?: boolean;
  mode?: string;
  controller_address?: string;
  controller_secret?: string;
}): Promise<KernelInstance> {
  return apiRequest<KernelInstance>(
    '/api/v1/admin/clash/runtime/start',
    {
      method: 'POST',
      body: payload,
    },
    decoder<KernelInstance>(),
  );
}

export async function stopKernel(
  kernelType = 'mihomo',
): Promise<{ stopped: boolean }> {
  return apiRequest<{ stopped: boolean }>(
    '/api/v1/admin/clash/runtime/stop',
    {
      method: 'POST',
      body: { kernel_type: kernelType },
    },
    decoder<{ stopped: boolean }>(),
  );
}

export async function reloadKernel(payload: {
  kernel_type?: string;
  allow_lan?: boolean;
  mode?: string;
  controller_address?: string;
  controller_secret?: string;
}): Promise<KernelInstance> {
  return apiRequest<KernelInstance>(
    '/api/v1/admin/clash/runtime/reload',
    {
      method: 'POST',
      body: payload,
    },
    decoder<KernelInstance>(),
  );
}

export async function getActiveRuntimeConfig(
  kernelType = 'mihomo',
): Promise<RenderResult> {
  return apiRequest<RenderResult>(
    `/api/v1/admin/clash/runtime/config?kernel_type=${kernelType}`,
    {
      method: 'GET',
    },
    decoder<RenderResult>(),
  );
}

export async function getKernelLogs(
  kernelType = 'mihomo',
): Promise<{ logs: string[] }> {
  return apiRequest<{ logs: string[] }>(
    `/api/v1/admin/clash/runtime/logs?kernel_type=${kernelType}`,
    {
      method: 'GET',
    },
    decoder<{ logs: string[] }>(),
  );
}

export type InstalledKernelBinary = {
  kernel_type: string;
  install_path: string;
  binary_source: string;
  detected_version: string;
  file_name: string;
  release_tag?: string;
  installed_at: string;
};

export async function getKernelCapabilities(): Promise<{
  capabilities: KernelCapability[];
}> {
  return apiRequest<{ capabilities: KernelCapability[] }>(
    '/api/v1/admin/clash/kernels/capabilities',
    {
      method: 'GET',
    },
    decoder<{ capabilities: KernelCapability[] }>(),
  );
}

export async function inspectKernelBinary(
  installPath?: string,
): Promise<InstalledKernelBinary> {
  return apiRequest<InstalledKernelBinary>(
    '/api/v1/admin/clash/kernels/inspect',
    {
      method: 'POST',
      body: { install_path: installPath },
    },
    decoder<InstalledKernelBinary>(),
  );
}

export async function downloadKernelBinary(
  installPath?: string,
): Promise<InstalledKernelBinary> {
  return apiRequest<InstalledKernelBinary>(
    '/api/v1/admin/clash/kernels/download',
    {
      method: 'POST',
      body: { install_path: installPath },
    },
    decoder<InstalledKernelBinary>(),
  );
}

export async function uploadKernelBinary(
  file: File,
  installPath?: string,
): Promise<InstalledKernelBinary> {
  const formData = new FormData();
  formData.append('binary', file);
  if (installPath) {
    formData.append('install_path', installPath);
  }
  return apiRequest<InstalledKernelBinary>(
    '/api/v1/admin/clash/kernels/upload',
    {
      method: 'POST',
      body: formData,
    },
    decoder<InstalledKernelBinary>(),
  );
}

File: backend/internal/domain/clash/clash.go
package clash

import "time"

// SourceConfig Constants
const (
	SourceConfigStatusParsed              = "parsed"
	SourceConfigStatusImported            = "imported"
	SourceConfigSourceTypeUpload          = "upload"
	SourceConfigSourceTypeSubscriptionURL = "subscription_url"
)

// ProxyNode Constants
const (
	NodeTestStatusUnknown = "unknown"
	NodeTestStatusSuccess = "success"
	NodeTestStatusFailed  = "failed"
)

// PortProfile Strategy Constants
const (
	PortProfileStrategySelect      = "select"
	PortProfileStrategyURLTest     = "url-test"
	PortProfileStrategyFallback    = "fallback"
	PortProfileStrategyLoadBalance = "load-balance"
)

// KernelInstance Status Constants
const (
	KernelInstanceStatusStopped   = "stopped"
	KernelInstanceStatusStarting  = "starting"
	KernelInstanceStatusRunning   = "running"
	KernelInstanceStatusStopping  = "stopping"
	KernelInstanceStatusError     = "error"
	KernelInstanceStatusReloading = "reloading"
)

// SourceConfig represents a configuration file upload or subscription URL
type SourceConfig struct {
	ID             int        `json:"id"`
	SourceType     string     `json:"source_type"`
	SourceURL      string     `json:"source_url"`
	ContentType    string     `json:"content_type"`
	FetchedAt      *time.Time `json:"fetched_at"`
	Filename       string     `json:"filename"`
	ContentHash    string     `json:"content_hash"`
	RawContent     string     `json:"-"`
	Status         string     `json:"status"`
	TotalNodes     int        `json:"total_nodes"`
	ValidNodes     int        `json:"valid_nodes"`
	InvalidNodes   int        `json:"invalid_nodes"`
	DuplicateNodes int        `json:"duplicate_nodes"`
	ImportedNodes  int        `json:"imported_nodes"`
	UploadedBy     string     `json:"uploaded_by"`
	UploadedByID   int        `json:"uploaded_by_id"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
}

// ProxyNode represents a proxy node entity
type ProxyNode struct {
	ID               int        `json:"id"`
	SourceConfigID   int        `json:"source_config_id"`
	SourceConfigName string     `json:"source_config_name"`
	Name             string     `json:"name"`
	Type             string     `json:"type"`
	Server           string     `json:"server"`
	Port             int        `json:"port"`
	Tags             string     `json:"tags"`
	Fingerprint      string     `json:"-"`
	MetadataJSON     string     `json:"metadata_json"`
	Enabled          bool       `json:"enabled"`
	LastTestStatus   string     `json:"last_test_status"`
	LastLatencyMS    *int       `json:"last_latency_ms,omitempty"`
	LastTestError    string     `json:"last_test_error,omitempty"`
	LastTestedAt     *time.Time `json:"last_tested_at,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

// NodeTestResult represents a test attempt result
type NodeTestResult struct {
	ID           int       `json:"id"`
	NodeID       int       `json:"node_id"`
	TestType     string    `json:"test_type"`
	Success      bool      `json:"success"`
	LatencyMS    int       `json:"latency_ms"`
	ErrorMessage string    `json:"error_message"`
	TestedAt     time.Time `json:"tested_at"`
}

// PortProfileProxySettings holds proxy strategy options for a port profile
type PortProfileProxySettings struct {
	StrategyType          string `json:"strategy_type"`
	TestURL               string `json:"test_url"`
	TestIntervalSeconds   int    `json:"test_interval_seconds"`
	LoadBalanceStrategy   string `json:"load_balance_strategy"`
	LoadBalanceLazy       bool   `json:"load_balance_lazy"`
	LoadBalanceDisableUDP bool   `json:"load_balance_disable_udp"`
	UDPEnabled            bool   `json:"udp_enabled"`
	AuthEnabled           bool   `json:"auth_enabled"`
	AuthUsername          string `json:"auth_username"`
	AuthPassword          string `json:"auth_password"`
}

// DefaultPortProfileProxySettings returns sensible defaults
func DefaultPortProfileProxySettings() PortProfileProxySettings {
	return PortProfileProxySettings{
		StrategyType:        PortProfileStrategySelect,
		TestURL:             "https://cp.cloudflare.com/generate_204",
		TestIntervalSeconds: 300,
		LoadBalanceStrategy: "consistent-hashing",
		UDPEnabled:          true,
		AuthEnabled:         false,
	}
}

// PortProfile represents an inbound listening port profile
type PortProfile struct {
	ID               int                      `json:"id"`
	Name             string                   `json:"name"`
	ListenHost       string                   `json:"listen_host"`
	MixedPort        int                      `json:"mixed_port"`
	SocksPort        int                      `json:"socks_port"`
	HTTPPort         int                      `json:"http_port"`
	ProxySettings    PortProfileProxySettings `json:"proxy_settings"`
	IncludeInRuntime bool                     `json:"include_in_runtime"`
	KernelType       string                   `json:"kernel_type"`
	CreatedAt        time.Time                `json:"created_at"`
	UpdatedAt        time.Time                `json:"updated_at"`
}

// PortProfileTemplate represents a reusable template for creating port profiles
type PortProfileTemplate struct {
	ID            int                      `json:"id"`
	Name          string                   `json:"name"`
	Description   string                   `json:"description"`
	MixedPort     int                      `json:"mixed_port"`
	ProxySettings PortProfileProxySettings `json:"proxy_settings"`
	CreatedAt     time.Time                `json:"created_at"`
	UpdatedAt     time.Time                `json:"updated_at"`
}

// PortProfileNode represents the binding between a port profile and a node (with ID & fingerprint persistence)
type PortProfileNode struct {
	ID              int       `json:"id"`
	PortProfileID   int       `json:"port_profile_id"`
	ProxyNodeID     int       `json:"proxy_node_id"`
	NodeFingerprint string    `json:"node_fingerprint"`
	SortOrder       int       `json:"sort_order"`
	CreatedAt       time.Time `json:"created_at"`
}

// RuntimeConfig holds the generated configuration snippet for a port profile
type RuntimeConfig struct {
	ID             int       `json:"id"`
	PortProfileID  int       `json:"port_profile_id"`
	KernelType     string    `json:"kernel_type"`
	Checksum       string    `json:"checksum"`
	RenderedConfig string    `json:"rendered_config"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

// KernelInstance represents the runtime process & state of Mihomo kernel
type KernelInstance struct {
	ID                   int        `json:"id"`
	KernelType           string     `json:"kernel_type"`
	Status               string     `json:"status"`
	PID                  *int       `json:"pid,omitempty"`
	WorkDir              string     `json:"work_dir"`
	ConfigPath           string     `json:"config_path"`
	ControllerAddress    string     `json:"controller_address"`
	ControllerSecret     string     `json:"-"`
	ActiveConfigChecksum string     `json:"active_config_checksum"`
	ActiveProfileCount   int        `json:"active_profile_count"`
	ActiveListenerCount  int        `json:"active_listener_count"`
	LastAction           string     `json:"last_action"`
	LastError            string     `json:"last_error"`
	LastStartedAt        *time.Time `json:"last_started_at,omitempty"`
	LastStoppedAt        *time.Time `json:"last_stopped_at,omitempty"`
	LastReloadedAt       *time.Time `json:"last_reloaded_at,omitempty"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
}

// KernelCapability describes the features supported by a kernel
type KernelCapability struct {
	KernelType          string   `json:"kernel_type"`
	SupportsReload      bool     `json:"supports_reload"`
	SupportsExternalAPI bool     `json:"supports_external_api"`
	SupportsHealthCheck bool     `json:"supports_health_check"`
	SupportedStrategies []string `json:"supported_strategies"`
}

// ProxyNodeFilter defines query parameters for node listing
type ProxyNodeFilter struct {
	Keyword        string
	SourceConfigID int
	Enabled        *bool
}
